### ISSUE TITLE
get contour polygon and fill it with color

### ISSUE THREAD

**Issue by xunsy** [uber/deck.gl#2457]
#### Background 
hi  ! i'm a student from china .I think you're doing a great job!  when i use it , i find something that might be usefuli to others . so i want to make a contribution for it and i tryed to pull requests  but  i don't  kown how to do it . so i submit  this issue for  what  i did .
when  i used contourlayer ,i want to get result like this 
![20181123174210](https://user-images.githubusercontent.com/22497642/48936884-44c0c280-ef47-11e8-8374-5d3ff68f52f2.png)
 

so i did something on it . in marching-squares.js ,i only change the CODE_OFFSET_MAP ,   so that the contourSegments is very  easy to be  connect  .  because  the contourSegments coordinate  will from [a0,a1], [a2,a1] become to [a0,a1], [a1,a2]
 then it's easy to make polygon out of contourSegments  ;
the picture is the result  of this way . and the code in my forks ;  hope  my job  can useful.  thanks for reading it .

**Comment by 1chandu** [comment]
@xunsy thank you for you comments.
For filled polygons, marching squares algorithm proposes a separate algorithm which uses 3 bit code for each corner of the grid cell : check `Isoband` section in https://en.wikipedia.org/wiki/Marching_squares.

This is currently being under development, I have couple of PRs: #2429  #2434 

Regarding the approach you suggested, I see following issues

1.  It doesn't handle the case when a contour begins and ends on the edge. To complete the polygon you will have to add line segment along the edge.
2. Same as above, but gets even complicated when contour begins on one edge and ends on another edge, you need to now know which corners (1 or 2) are part of the contour boundary.
3. Most filled contour use cases, would specify lower and higher bound, where you have two sets of contour lines, which is also not handled this way.
4. This is not suitable to do on GPU, where order of line segments is not controlled. (We plan to address this using above mentioned algorithm).

There might be other issues, but I haven't digged deep into your code.

Can you checkout this branch : https://github.com/uber/deck.gl/tree/IsoBands-ContourLayer and see if  this servers your purpose?  Documentation here : https://github.com/uber/deck.gl/blob/IsoBands-ContourLayer/docs/layers/contour-layer.md
Basically to get filled polygon you would specify threshold as array of two values like below, you can also specify [0, 6] as your threshold.

```
  const layer = new ContourLayer({
    id: 'contourLayer',
    contours: [
      {threshold: [6, 10], color: [0, 0, 255]} // => Isoband for threshold range [6, 10)
    ],
    cellSize: 200,
    getPosition: d => d.COORDINATES,
  });
```
Highly appreciate any feedback and comments.

**Comment by xunsy** [comment]
i'm so excited  when i get the reply. after careful thought, i think the issues 1,2,3 i have solved it . i drew a picture  to show  what  my code  did ; The following picture may  fits your situation which refer to issues 1,2,3.
![example](https://user-images.githubusercontent.com/22497642/49080860-f5003500-f27f-11e8-9839-0fc98c543d01.png)
i labeled x axis ,y axis and weight in cell;  the **Isoline**  will not get the results that the picture shows;
 on the contrary  ,you will get the results  like the following  picture;
![example2](https://user-images.githubusercontent.com/22497642/49081253-3513e780-f281-11e8-90ac-c1ea6411b93d.png)  
I marked out the difference between the two pictures with purple circles.  the reason why  leads to this is that the Marching squares algorithm is not begin at [0,0]  but [-1,-1];    it means  the weight in  cell  will be like this following picture: 
![example3](https://user-images.githubusercontent.com/22497642/49081749-b8820880-f282-11e8-903d-4f44ccada000.png)  you can see  there is an extra lap of data outside the grid ;

for issues 2 ,i use  the  code  ,which  changed the object of CODE_OFFSET_MAP ,the  coordinate  we get will like [a0,a1][a2,a3][a1,a2][a3,a0] ;It's been tested that all segments are end to end; this is the schematic diagram;and the red arrows is direction;
![example4](https://user-images.githubusercontent.com/22497642/49083647-2f6dd000-f288-11e8-9591-f73a34db902d.png)

for issues 3  ,  my idea is  make polygon out of contourSegments in the function of  thresholds.forEach.
so No two segments will have same coordinate  [i,j]

the following picture is my result 
![20181123174210](https://user-images.githubusercontent.com/22497642/49084985-eae43380-f28b-11e8-92d8-e4108ccddf68.png)  and  the code is[ here](https://github.com/xunsy/deck.gl/pull/2)


thanks  for you reading and Welcome To Suggest Improvements.

**Comment by 1chandu** [comment]
> for issues 3 , my idea is make polygon out of contourSegments in the function of thresholds.forEach.
so No two segments will have same coordinate [i,j]

sorry, I didn't quite understand this. When a range is specified, lets say [5, 10], I am assuming a contour is generated with value 5 and another one with value 10. So you have two sets of polygons now, how will you just render the are between these two contour lines?  Remember result of each contour could be multiple closed polygons.

The code you pointed is hard to read as there are multiple changes, as mentioned earlier, this approach has few issues and also not suitable for GPU acceleration. You can check the PR mentioned above to see if that serves your needs. 

Regarding the code, you can update your code to latest master branch and add a new layer (and a new set of utility methods) under `modules/experimental-layers`, we will be happy to take a look at your PR.

**Comment by xunsy** [comment]
Thank you for your patience!
Regarding the code:
first:  assume  i want to get a contour is generated with value 5 and another one with value 10; so i will get all contour  segments  are  generated with value 5 first and They make up polygons .Let's say that the set of polygons is **A**;then  i will get all contour  segments  are  generated with value 10 and They make up polygons ;Let's say that the set of polygons is **B**; then we merge sets A and B;

finally :
i will check the PR  mentioned above;
 and  try   to new PR ;
thanks for you answer;

**Comment by 1chandu** [comment]
>  then we merge sets A and B;

This will not give the correct result, which is we should be filling only the region who's values are in [5, 10] range. Depending on how the weight is spread in the texture, A (when rendered as you mentioned) will fill the area either below are above value `5`, and same way with B.  You can also experiment this with two texture 1. Value at center is high (lets say 50) and it reduces it value as it goes away from the center, 2. Value at center is low (lets say 0) and it increase as it goes away from the center.

**Comment by 1chandu** [comment]
Closing due to inactivity , please re-open for any un-resolved items.
